feat: add cookie-based session affinity for Istio

munday-tech opened the pull request:
Last year a change was made to remove `X-Chainlit-Session-id` header from websocket connections, causing kubernetes clusters using istio to lose the ability to maintain sticky sessions. Discussed here: https://github.com/Chainlit/chainlit/issues/1892

This PR adds new functionality to allow users to enable the addition of the users session id as a cookie, that istio can use for sticky sessions. This simply sets a cookie on the client-side when a websocket connection occurs.

Changes:
- Added `enable_session_cookie` to enable this option under project settings. Defaults to off.
- Added a FastAPI endpoint to set session cookie
- Added websocket command to trigger the above API

willydouhard commented:
This would be useful for everybody, I don't think we need to make it opt-in.
Also, instead of relying on an extra websocket message to get the cookie, we could make the call directly [here](https://github.com/Chainlit/chainlit/blob/main/libs/react-client/src/useChatSession.ts#L108)

willydouhard commented on backend/chainlit/server.py:
```diff
@@ -302,6 +299,25 @@ async def teams_endpoint(req: Request):
 
 
+@app.get("/set-session-cookie")
+async def set_session_cookie(session_id: str, response: Response, request: Request):
```
an auth check would be welcomed here, example available [here](set_auth_cookie)

munday-tech replied:
This link is dead :(

willydouhard replied:
here you go https://github.com/Chainlit/chainlit/blob/main/backend/chainlit/server.py#L1243

willydouhard commented:
Are you willing to implement the changes @munday-tech? I think the whole community would benefit from this

munday-tech commented:
Can do it this week @willydouhard.

munday-tech commented:
@willydouhard Made the changes.

1. Removed config option, happens every websocket connect now
2. Updated the set cookie to `POST` from `GET`
3. Added in the auth check as requested
4. Added in test cases

The `POST` happens in the `.connect()` of the websocket as it has to occur after the websocket has been created (So we can fetch session id and map to user for auth)

My ts/react isn't great so let me know if anything needs to be improved :)

munday-tech commented:
@willydouhard any plans for when we can review / get this one in :)?

willydouhard reviewed (approved):
lgtm

willydouhard commented:
I will merge this and make two small modifications:

1. Use the chainlit api client to make the call
2. Move the set cookie before the connection and not once the connection is successful (since the session is created on the server during the connection and we need stickiness right from there)

Thank you for your contribution!
